server/entity/arrow.go

```diff
@@ -0,0 +1,285 @@
+package entity
+
+import (
+	"github.com/df-mc/dragonfly/server/block/cube"
+	"github.com/df-mc/dragonfly/server/entity/action"
+	"github.com/df-mc/dragonfly/server/entity/damage"
+	"github.com/df-mc/dragonfly/server/entity/physics"
+	"github.com/df-mc/dragonfly/server/entity/physics/trace"
+	"github.com/df-mc/dragonfly/server/internal/nbtconv"
+	"github.com/df-mc/dragonfly/server/item"
+	"github.com/df-mc/dragonfly/server/item/potion"
+	"github.com/df-mc/dragonfly/server/world"
+	"github.com/df-mc/dragonfly/server/world/sound"
+	"github.com/go-gl/mathgl/mgl64"
+	"math"
+	"math/rand"
+	"time"
+)
+
+// Arrow is used as ammunition for bows, crossbows, and dispensers. Arrows can be modified to
+// imbue status effects on players and mobs.
+type Arrow struct {
+	transform
+	yaw, pitch float64
+	baseDamage float64
+
+	ticksLived, collisionTicks int
+
+	collidedBlockPos cube.Pos
+	collidedBlock    world.Block
+
+	tip potion.Potion
+
+	closeNextTick, critical bool
+
+	owner                        world.Entity
+	shotByPlayer, shotInCreative bool
+
+	c *ProjectileComputer
+}
+
+// NewArrow ...
+func NewArrow(pos mgl64.Vec3, yaw, pitch float64, owner world.Entity, critical, shotByPlayer, shotInCreative bool, baseDamage float64, tip potion.Potion) *Arrow {
```

Comment:
This seems like a ton of fields crammed into one function. First of all I'd move the `tip` parameter to a separate `NewTippedArrow()` function.

Reply:
I would also rename shotInCreative to something that indicates its actual functionality: Not receiving an item when picking it up.

Reply:
Same for `shotByPlayer`, would rename that to a name that implies it cannot be picked up.

Reply:
The thing is, `shotInCreative` only does not allow you to receive an item when picking it up when the player's gamemode is not creative. Personally I feel like `shotByPlayer` & `shotByCreative` make the most sense for what they do: control specific pickup options based on the player and their gamemode?

Reply:
Also, when I do move over `tip` to it's own `NewTippedArrow()` function, how should that look in the means of the projectile initialization? Should the `New` function switch over to use `NewTippedArrow()` as well, or shall I create another function for `NewTipped()` which does the same thing as `New` but specifically for tipped arrows?

Reply:
> Personally I feel like shotByPlayer & shotByCreative make the most sense for what they do: control specific pickup options based on the player and their gamemode?
The name's "shotByPlayer" and "shotByCreative" does not make it obvious at all what they actually do. And although their behaviors are only used in vanilla when a player shoots it and/or is in creative, it does not mean that the behaviors would like to be used by developers when making things such as mini-games

Reply:
Just refactored both of those variables, but I'm still not sure on the `NewTippedArrow()` discussion. Could I have some thoughts on what I mentioned above?

Reply:
I'd have a `NewTippedArrow` which accepts a `tip` parameter, then a `NewArrow` function which does not and calls NewTippedArrow with an 'empty' potion effect.

Reply:
Should `New` use `NewTippedArrow` then?